File: CCC 2020/CCC_2020_q5.py
from math import ceil
from collections import deque
from time import time

def coords(n, M, N):
    fact = set()
    for i in range(1, ceil(n**0.5)+1):
        if not n%i:
            b = n//i
            if i <= M and b <= N:
                fact.add((i-1, b-1))
            if b <= M and i <= N:
                fact.add((b-1, i-1))
    return deque(fact)

def escape(M, N, grid):
    pos = (0, 0)
    seen = {}
    junctions = deque()
    factors = {}

    if grid[0][0] == M*N:
        return 'yes'

    while True:
        
        num = grid[pos[0]][pos[1]]
        try:
            options = factors[num]
        except:
            options = coords(num, M, N)
            factors[num] = options
        
        seen[pos] = None

        try:
            pos = options.popleft()
            if not junctions and pos == (0, 0):
                return 'no'
            if options:
                junctions.append(options)
        except:
            pass

        try:
            seen[pos]
            if not junctions:
                return 'no'

            try:
                options = junctions.popleft()
                pos = options.popleft()
                if options:
                    junctions.append(options)
            except:
                pass
        except:
            pass
        
        if pos == (M-1, N-1):
            return 'yes'

##
##M = int(input())
##N = int(input())
##grid = deque(tuple(map(int, input().split())) for i in range(M))
##print(escape(M, N, grid))

test={1: range(1, 8), 2: range(1, 7), 3: range(1, 6),
      5: range(1, 5), 6: range(1, 7), 7: range(1, 8)}
##for i in test.keys():
##    for j in test[i]:
##        f = open(f'j5.0{i}.0{j}.in', 'r')
##        f1 = open(f'j5.0{i}.0{j}.out', 'r')
##        case = list([int(f.readline()), int(f.readline())])
##        case.append(tuple(tuple(map(int, f.readline().split())) for i in range(case[0])))
##        print(f'j5.0{i}.0{j}.in')
##        t = time()
##        result = escape(case[0], case[1], case[2])
##        print(time()-t)
##        ans = f1.read().strip()
##        print(ans in result)
##        #print(result)
##        print()
##        f.close(), f1.close()
f = open(f'j5.07.05.in', 'r')
f1 = open(f'j5.07.05.out', 'r')
case = list([int(f.readline()), int(f.readline())])
case.append(tuple(tuple(map(int, f.readline().split())) for i in range(case[0])))
t = time()
result = escape(case[0], case[1], case[2])
print(time()-t)
ans = f1.read().strip()
print(ans in result)
print()
f.close(), f1.close()

File: CCC 2020/CCC_2020_q5.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
#include <math.h>
#include <deque>
using namespace std;

deque<vector<int>> coords(int n, int M, int N){
    deque<vector<int>> fact;
    for (int i=1; i<ceil(sqrt(n))+1; i++){
        if (!(n%i)){
            int b = n/i;
            if (i <= M && b <= N){
                vector<int> temp{i-1, b-1};
                if (find(fact.begin(), fact.end(), temp) == fact.end()){
                    fact.push_back(temp);
                }
            }
            
            if (b <= M and i <= N){
                vector<int> temp{b-1, i-1};
                if (find(fact.begin(), fact.end(), temp) == fact.end()){
                    fact.push_back(temp);
                }
            }
        }
    }
    
    return fact;
}

string escape(int m, int n, vector<vector<int>> grid){
    vector<int> pos{0, 0};
    vector<vector<int>> seen;
    deque<deque<vector<int>>> junctions;
    map<int, deque<vector<int>>> factors;

    if (grid[0][0] == m*n){
        return "yes";
    }

    while (true){
        int num = grid[pos[0]][pos[1]];
        deque<vector<int>> options;

        if (factors.find(num) == factors.end()){
            options = coords(num, m, n);
            factors[num] = options;
        }
        else{
            options = factors[num];
        }
        
        seen.push_back(pos);

        if (!options.empty()){
            pos = options.front();
            options.pop_front();
            
            if (junctions.empty() && pos[0] == 0 && pos[1] == 0){
                return "no";
            }
            
            if (!options.empty()){
                junctions.push_back(options);
            }
        }

        if (find(seen.begin(), seen.end(), pos) != seen.end()){
            if (junctions.empty()){
                return "no";
            }
            if (!options.empty() &&  !junctions.empty()){
                options = junctions.front();
                junctions.pop_front();
                pos = options.front();
                options.pop_front();
                if (!options.empty()){
                    junctions.push_back(options);
                }
            }
        }

        if (pos[0] == m-1 && pos[1] == n-1){
            return "yes";
        }
    }
}

int main(){
    int M, N;
    vector<vector<int>> grid;
    
    cin>>M>>N;
    grid.resize(M);
    
    for (int i=0; i<M; i++){
        grid[i].resize(N);
        for (int j=0; j<N; j++){
            cin>>grid[i][j];
        }
    }
    
    cout<<escape(M, N, grid)<<endl;
    return 0;
}
